### PR TITLE
DDP-4222 new event types for age-up

### PR DESCRIPTION
**Pull request by yufengwng**
## Context and the big picture

This PR implements the new event trigger/action types we'll need to support Age Up. These changes should unblock a number of other tickets. See associated ticket and design doc for more details.

Some aspects are not implemented yet, in particular:
* Handling of "hidden" activity instances. Will require changes to the API routes.
* Handling of notification actions with `INVITATION_EMAIL` type. Will require changes to `PubSubMessageBuilder` to build the email appropriately.
* Updating email notification action to add invitation guid as a substitution value and overriding recipient email.

I will implement these in separate PRs.

Some tips on reading this PR:
1. Start with the liquibase XML files to see the database changes.
2. See `EventTriggerDao` and `EventActionDao` for how we insert these new event types into database.
3. Then checkout `EventConfigurationDto`, `EventConfiguration`, `EventDao` and `EventDao.sql.stg` for how we query them out.
4. Go through each of the `*EventAction` classes, their tests, and related files to see how they work.
5. Lastly, checkout the study-builder changes to see how these are configured by a study.
 
 ## What type of change is this?
  
  - [ ] Bugfix (non-breaking change which fixes an issue)
  - [x] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [ ] Experimental proof-of-concept
  - [ ] Infosec/Compliance remediation
  
 ## Risk Assessment
  - [ ] These changes are **HIGH RISK**.
    - [ ] There is an elevated risk that may impact a large number of features
    - [ ] There is an elevated risk to security and privacy
    - [ ] These changes introduce new 3rd party dependencies, alter the network perimeter, or contain
    major upgrades to persistent storage or SaaS dependencies such as auth0, sendgrid, or easypost    
  - [x] These changes are low risk, do not impact security/privacy, do not impact pepper shared
  components/modules, and have no major upgrades to 3rd party services or libraries
 
 ### FUD Score 

  - [x] :relaxed: All good, business as usual!
  - [ ] :sweat_smile: There might be some issues here
  - [ ] :scream:I'm sweaty and nervous
 
 ## How to we demo these changes?
 
 - [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
 - [ ] Getting dev into a state where this is user-visible requires some tech fiddling.  I have
 documented these steps in the related ticket.
 - [x] Requires other features before it's human visible. Will need study configurations to drive these.
 - [ ] I have no idea how to demo this.  Please help me!
 
 ## UI/UX
  - [x] There ain't no UI/UX in these changes
  - [ ] My changes have passed muster with Erin and Jen via an over-the-shoulder review, screenshots, etc.
  - [ ] Erin and Jen have not had an opportunity to provide feedback yet
 
###  Backend
   - [x] I have considered error handling and notification to slack alert rooms via `LOG.error()`.
     #### Route Changes
     - [ ] My changes involve a new route changes to an existing route, so the who/what/when are logged using the usual logback mechanism
at the _start_ of the route, prior to any validation, as well as at the end of the route when work completes.
     ### Housekeeping Changes
     - [ ] My changes involve updates to housekeeping, and I am logging the who/what/when via logback.
  ### Client-side
   - [ ] My changes do not involve backend route change, but I have useful logging statements nonetheless.

## Analytics
 - [ ] I have discussed the analytics needs at both a platform and study level with Jen and instrumented code
 accordingly.
 - [x] There are no analytics requirements for these changes

###  Backend
   - [ ] These changes alter fundamental auth filter logic
   - [ ] These changes introduce a new backend API route that is behind an auth filter
   - [ ] These changes introduce a new backend API that _does not require auth_
   - [ ] These changes alter auth0 internals (rules, hosted login, configuration, etc.)
   - [ ] These changes expose new information to elastic and I have verified that PHI is not being exposed to the public
    
  ### Browser
  - [ ] These changes alter what we store in browser storage
  
  ### Mobile
  - [ ] TBD 
 
 ## Performance and Stability
  - [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, and performance bottlenecks and written a brief summary in this PR
  - [ ] I'm unsure about stability, performance, fault tolerance, and graceful degradation.  Please help!
  
## Testing
 - [x] I have written automated positive tests
 - [x] I have written automated negative tests
 - [ ] I have written zero automated tests but have poked around locally to verify proper functionality
 - [ ] The jira ticket has acceptance criteria and Kiara has in the information she needs to test my changes
 
## Release
 - [x] These changes require no special release procedures--just code!
 - [ ] Releasing these changes requires special handling
   - [ ] I have documented the special procedures in the release plan document
   - [ ] The special procedures require minimal downtime
   - [ ] The special procedures require unavoidable, extended downtime for:
     - [ ] Participant UX
     - [ ] Study staff

### CODE REVIEW

**Comment by yufengwng** [comment]
please run tests

**Comment by yufengwng** [comment]
please run tests

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/ActivityInstanceDao.java [hunk 1]
```diff
@@ -340,6 +340,8 @@ public Long getActivityInstanceIdByGuid(Handle handle, String guid) {
                             createdAtMillis, statusTypeCode, isActivityWriteOnce, isActivityInstanceReadonly);
                     long createdAt = rs.getLong(SqlConstants.ActivityInstanceTable.CREATED_AT);
                     boolean isFollowup = rs.getBoolean(SqlConstants.StudyActivityTable.IS_FOLLOWUP);
+                    boolean isHidden = rs.getBoolean(SqlConstants.ActivityInstanceTable.IS_HIDDEN);
+                    // todo: handle isHidden
```
Comment (spookymotion) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/ActivityInstanceDao.java:344]: Is this todo relevant anymore?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/FormInstanceDao.java [hunk 2]
```diff
@@ -114,6 +114,8 @@ public FormInstance getBaseFormByGuid(Handle handle, String instanceGuid, String
                 Long lastUpdatedTemplateId = (Long)rs.getObject(FormActivitySettingTable.LAST_UPDATED_TEXT_TEMPLATE_ID);
                 LocalDateTime lastUpdated = rs.getObject(FormActivitySettingTable.LAST_UPDATED, LocalDateTime.class);
                 boolean isFollowup = rs.getBoolean(StudyActivityTable.IS_FOLLOWUP);
+                boolean isHidden = rs.getBoolean(ActivityInstanceTable.IS_HIDDEN);
+                // todo: handle isHidden
```
Comment (spookymotion) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/FormInstanceDao.java:118]: ^

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/EventActionDao.java [hunk 3]
```diff
@@ -49,16 +51,22 @@
...
+    @CreateSqlObject
+    EventActionSql getEventActionSql();
+
+    default long insertInvitationEmailNotificationAction(SendgridEmailEventActionDto sendgridTemplateDto) {
+        return insertNotificationAction(sendgridTemplateDto, NotificationType.INVITATION_EMAIL);
+    }
 
     default long insertStudyNotificationAction(SendgridEmailEventActionDto sendgridTemplateDto) {
-        return _insertNotificationAction(sendgridTemplateDto, NotificationType.STUDY_EMAIL);
+        return insertNotificationAction(sendgridTemplateDto, NotificationType.STUDY_EMAIL);
```
Comment (spookymotion) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/EventActionDao.java:62]: Are these now entry points to the class?

File: study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/EventBuilder.java [hunk 4]
```diff
@@ -250,7 +256,17 @@ private String actionAsStr(Config actionCfg) {
         } else if (EventActionType.COPY_ANSWER.name().equals(type)) {
             String copySourceQuestionStableId = actionCfg.getString("copySourceQuestionStableId");
             CopyAnswerTarget copyTarget = actionCfg.getEnum(CopyAnswerTarget.class, "copyTarget");
-            return String.format("COPY_ANSWER/%s/%s", copySourceQuestionStableId, copyTarget);
+            return String.format("%s/%s/%s", type, copySourceQuestionStableId, copyTarget);
```
Comment (spookymotion) [inline study-builder/src/main/java/org/broadinstitute/ddp/studybuilder/EventBuilder.java:259]: I understand the code was like this when you got here, but the team should pick a string format solution and use it all over the place. I think format(%s) is not used as much. Personally speaking, it can be hard to read, because you have to line up the var args with the format tokens.

**Review by spookymotion — DISMISSED** [review]

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/ActivityInstanceDao.java (hunk 1 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/ActivityInstanceDao.java:344]: It's fine to leave this here -- I have removed it in the follow-up PRs.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/EventActionDao.java (hunk 3 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/EventActionDao.java:62]: I guess so -- if you need different types of notifcation, you can use the appropriate methods. This `insertNotificationAction()` method wasn't used anywhere else anyways, so I made it private.

File: pepper-apis/src/main/resources/db-changes/schema/DDP-4222-age-up-event-actions.xml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                             foreignKeyName="event_action_target_activity_activity_fk"/>
+            </column>
+        </createTable>
+        <addUniqueConstraint tableName="event_action_target_activity" columnNames="event_action_id, activity_id"
+                             constraintName="event_action_target_activity_action_activity_uk"/>
+    </changeSet>
+
+    <changeSet author="yufeng" id="20200102-add-is-hidden-column-to-activity-instance">
+        <addColumn tableName="activity_instance">
+            <column name="is_hidden" type="boolean"/>
```
Comment (ssettipalli) [inline pepper-apis/src/main/resources/db-changes/schema/DDP-4222-age-up-event-actions.xml]: wondering if it's handy to set default value "false" considering most of the activity instances are not hidden.

Comment (yufengwng) [reply pepper-apis/src/main/resources/db-changes/schema/DDP-4222-age-up-event-actions.xml]: I had considered it. But I wanted to make it non-null, and I couldn't quite figure out how to declare all of that with liquibase. Let me try again.

**Comment by yufengwng** [comment]
please run tests

File: pepper-apis/src/main/resources/db-changes/schema/DDP-4222-age-up-event-actions.xml (hunk 5 above)
Comment (yufengwng) [reply pepper-apis/src/main/resources/db-changes/schema/DDP-4222-age-up-event-actions.xml]: Okay I think I figured out the trick.

**Review by ssettipalli — APPROVED** [review]

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/event/CreateInvitationEventAction.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    @Override
+    public void doAction(PexInterpreter interpreter, Handle handle, EventSignal signal) {
+        InvitationCreatedSignal nextSignal = run(handle, signal);
+        EventService.getInstance().processAllActionsForEventSignal(handle, nextSignal);
+    }
+
+    InvitationCreatedSignal run(Handle handle, EventSignal signal) {
+        String contactEmail = fetchContactEmail(handle, signal);
+        if (!MiscUtil.isEmailFormatValid(contactEmail)) {
+            throw new DDPException("Answer contact email is not a valid email");
```
Comment (zyme) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/event/CreateInvitationEventAction.java:50]: Maybe include `contactEmail` in the exception text?

Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/model/event/CreateInvitationEventAction.java:50]: Will do in one of the follow-up PRs.
